pkg/tempo/servicegraphprocessor/processor.go

```diff
@@ -0,0 +1,255 @@
+		p.serviceGraphUntaggedSpansTotal,
+	}
+
+	for _, c := range cs {
+		if err := p.reg.Register(c); err != nil {
+			return err
+		}
+	}
+
+	// Collect unpaired spans when evicting items from the store during
+	// periodic cleanup
+	p.store.OnEvicted(func(s string, i interface{}) {
+		e := i.(edge)
+		if !e.complete() {
+			p.serviceGraphUnpairedSpansTotal.WithLabelValues(e.clientService, e.serverService).Inc()
+		}
+	})
+
+	return nil
+}
+
+func (p *processor) Shutdown(context.Context) error {
+	p.unregisterMetrics()
+	return nil
+}
+
+func (p *processor) unregisterMetrics() {
+	cs := []prometheus.Collector{
+		p.serviceGraphRequestTotal,
+		p.serviceGraphRequestFailedTotal,
+		p.serviceGraphRequestHistogram,
+		p.serviceGraphUnpairedSpansTotal,
+		p.serviceGraphUntaggedSpansTotal,
+	}
+
+	for _, c := range cs {
+		p.reg.Unregister(c)
+	}
+}
+
+func (p *processor) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{}
+}
+
+func (p *processor) ConsumeTraces(ctx context.Context, td pdata.Traces) error {
+	level.Debug(p.logger).Log("msg", "consuming traces")
+
+	var errs error
+	for _, trace := range batchpersignal.SplitTraces(td) {
+		if err := p.consume(trace); err != nil {
+			if errors.Is(err, errTooManyItems) {
+				level.Warn(p.logger).Log("msg", "skipped processing of spans", "maxItems", p.maxItems, "err", errTooManyItems)
+				break
+			}
+			errs = multierror.Append(errs, err)
+		}
+	}
+	if errs != nil {
+		level.Error(p.logger).Log("msg", "failed consuming traces", "err", errs)
+	}
+
+	p.collectMetrics()
+
+	return p.nextConsumer.ConsumeTraces(ctx, td)
+}
+
+func (p *processor) collectMetrics() {
+	for k, v := range p.store.Items() {
+		e := v.Object.(edge)
+		if e.complete() {
+			p.serviceGraphRequestTotal.WithLabelValues(e.clientService, e.serverService).Inc()
+			if e.failed {
+				p.serviceGraphRequestFailedTotal.WithLabelValues(e.clientService, e.serverService).Inc()
+			}
+			p.serviceGraphRequestHistogram.WithLabelValues(e.clientService, e.serverService).Observe(e.serverLatency.Seconds())
+			p.store.Delete(k)
+		}
+	}
+}
+
+func (p *processor) consume(trace pdata.Traces) error {
+	rSpansSlice := trace.ResourceSpans()
+	for i := 0; i < rSpansSlice.Len(); i++ {
+		rSpan := rSpansSlice.At(i)
+
+		svc, ok := rSpan.Resource().Attributes().Get(conventions.AttributeServiceName)
+		if !ok {
+			continue
+		}
+
+		ilsSlice := rSpan.InstrumentationLibrarySpans()
+		for j := 0; j < ilsSlice.Len(); j++ {
+			ils := ilsSlice.At(j)
+
+			for k := 0; k < ils.Spans().Len(); k++ {
+				span := ils.Spans().At(k)
+
+				if p.store.ItemCount() >= p.maxItems {
+					return errTooManyItems
+				}
+
+				switch span.Kind() {
+				case pdata.SpanKindClient:
+					k := key(span.TraceID().HexString(), span.SpanID().HexString())
+
+					var e edge
+					if v, ok := p.store.Get(k); ok {
+						e = v.(edge)
+					}
+					e.clientService = svc.StringVal()
+					e.clientLatency = spanDuration(span)
+					p.store.SetDefault(k, e)
```

Comment:
I might be confused about how we are collecting our latency data... but if I understand this correctly:
- for every span, we update the store with a new latency for that service
- once per `ConsumeTraces` we observe this latency (in `collectMetrics`)

If we, for example, receive multiple spans with the same service name and durations `10s`, `5s` and `7s`. Am I right that we will only observe the latest duration (`7s`)? The previous durations will be replaced before we get to `collectMetrics`.

Second question about the latency: if a service emits multiple spans for a single request there might be one long span with a bunch of short child spans. Is the latency equal to the longest span or to the average of all spans?

Reply:
Aggregation is done by the parent (client) span, so unless two child (server) spans share the same parent span, the problem should not occur. I understand that should not happen if instrumentation is done correctly, but I will check what the specification says.

Reply:
Oh okay, I missed that we only gather latencies from client and server spans. Other spans will be internal and thus ignored 🤦🏻  The design makes way more sense now 😅 

A single parent span might still create multiple child spans, but I think in general it's a good practice to create one span per outgoing request.
We could log a warning if there is already a latency set, but I'm worried this could flood the logs.